keps/sig-storage/0035-20190116-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,140 @@
+---
+kep-number: 35
+title: ExecutionHook 
+authors:
+  - "@jingxu97"
+  - "@xing-yang"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+  - sig-node
+  - sig-architecture
+reviewers:
+  - "@saad-ali"
+  - "@thockin"
+approvers:
+  - "@thockin"
+  - "@saad-ali"
+editor: TBD
+creation-date: 2019-1-16
+last-updated: 2019-1-16
+status: implementable
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+ExecutionHook API change
+
+## Table of Contents
+
+  * [Title](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories](#user-stories)
+      * [Workarounds](#workarounds)
+      * [Alternatives](#alternatives)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Implementation History](#implementation-history)
+
+## Summary
+
+Volume snapshot support was introduced in Kubernetes v1.12 as an alpha feature. 
+In the alpha implementation of snapshots for Kubernetes, there is no snapshot consistency
+guarantees beyond any guarantees provided by storage system (e.g. crash consistency).
+
+This proposal is aimed to address that limitation by providing an `ExecutionHook`
+in the `Container` struct. The snapshot controller will look up this hook before
+taking a snapshot and execute it accordingly.
+
+## Motivation
+
+The volume snapshot feature allows creating/deleting volume snapshots, and the
+ability to create new volumes from a snapshot natively using the Kubernetes API.
+
+However, application consistency is not guaranteed. An user has to figure out how
+to quiece an application before taking a snapshot and unquiece it after taking
+the snapshot.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container.
+Specifically it can be used to perform a quiese operation before a volume snapshot
+is taken and then perform an unquiesce operation after a volume snapshot is
+taken to resume the application.
+
+## Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+## Proposal
+
+An `ExecutionHook` is defined in the following.
+
+```
+// ExecutionHook defines a specific action that should be taken with retries and timeouts
+type ExecutionHook struct {
```

Comment:
What's the relationship with the existing container lifecycle hook and why need to introduce a new one?

Reply:
The lifecycle hook is called immediately after a container is created or immediately before a container is terminated.  The proposed execution hook is not tied to the start or termination time of the container.  Instead it is called before/after taking a snapshot while the container is still running.  So we can't use the lifecycle hook.  I'll add some clarification to the spec.

Reply:
Thanks for explanation!

While as we are introducing a new hook in vanilla container struct, it will be better if there're more general use cases/user stories instead of vol snapshot only.

Reply:
Other than Lifecycle hook, there is also Liveness Probe and Readiness Probe hooks in Container to check the health of a container.  So it seems that the existing hooks all have a specific purpose?  It is possible that a user may want to do some preparation work before an action and then do some post-processing after an action.  I am just not sure what those actions are, other than snapshotting.  Alternatively we can make this a more general-purpose pre-action and post-action hook.

Reply:
Yeah, it's probably there are more general purposes and it may need some thinking but would be great to be included in the user stories part as well. I am not against adding hooks for specific purpose of course, but then I would suggest to name it as sth more specific like VolSnapshotHook instead of ExecutionHook.

What we are trying to avoid is, we have a `ExecutionHook` which sounds like a general purpose hook, but oops, we are telling users that it can only be used for prep snapshot from design doc, code and all the way to user doc.

Reply:
See updated spec.